src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1371,24 +1371,30 @@ public override BoundNode VisitLocalDeclaration(BoundLocalDeclaration node)
 
             TypeWithAnnotations type = local.TypeWithAnnotations;
             TypeWithState valueType;
+            bool inferredType = node.InferredType;
             if (local.IsRef)
             {
                 valueType = VisitRefExpression(initializer, type);
             }
             else
             {
-                bool inferredType = node.InferredType;
-                valueType = VisitOptionalImplicitConversion(initializer, targetTypeOpt: inferredType ? default : type, useLegacyWarnings: true, trackMembers: true, AssignmentKind.Assignment);
-                if (inferredType)
+                valueType = VisitOptionalImplicitConversion(initializer, targetTypeOpt: inferredType ? default : type, useLegacyWarnings: true, trackMembers: true, AssignmentKind.Assignment);         
+            }
+
+            if (inferredType)
+            {
+                if (valueType.HasNullType)
                 {
-                    if (valueType.HasNullType)
-                    {
-                        Debug.Assert(type.Type.IsErrorType());
-                        valueType = type.ToTypeWithState();
-                    }
+                    Debug.Assert(type.Type.IsErrorType());
+                    valueType = type.ToTypeWithState();
+                }
 
-                    type = valueType.ToTypeWithAnnotations();
-                    _variableTypes[local] = type;
+                type = valueType.ToTypeWithAnnotations();
+                _variableTypes[local] = type;
+
+                if (node.DeclaredTypeOpt != null)
+                {
+                    SetAnalyzedNullability(node.DeclaredTypeOpt, new VisitResult(valueType, type), true);
```

Comment:
Please use `VisitTypeExpression`. `BoundTypeExpression`s can have child type expressions and this would not get them. Add a test with nested generics to confirm.

Reply:
Actually nevermind, we're in inferred state so it must be `var`, and that by definition doesn't have nested state.

---
In reply to: [290072374](https://github.com/dotnet/roslyn/pull/36078#discussion_r290072374) [](ancestors = 290072374)